# PopGenCode
Various bits of code for population genetic methods used in my publications

## Compute D<sub>xy</sub> from site frequency spectrum

Code used to compute D<sub>xy</sub> in 10kb-windows across the genome in Marques et al. 2018 in press.

### 1) Estimate 2D-SFS in windows using angsd

I used the code in [run10kdxy.pbs](http://github.com/marqueda/PopGenCode/blob/master/run10kdxy.pbs) to compute 2D-SFS in predefined windows given in a BED-format file (e.g. file windows.chrI.GLA.10k.bed listing all 10kb-windows on chromosome I). For each line in the BED-file, the script runs angsd twice to generate a temporary SAF file for the specific window for both populations (defined by bam file lists) and then runs realSFS, which computes the 2D-SFS. Chromsome, start and end position of the window and the realSFS output are then added to the outfile, one line per line (see file [example.10kwin.sfs](https://github.com/marqueda/PopGenCode/blob/master/run10kdxy.pbs)). Sorry for the hard-coded format...

### 2) Compute D<sub>xy</sub> from 2D-SFS file

Once you have a file generated by the above script with chromosome, start, end position and SFS entries (file [example.10kwin.sfs](http://github.com/marqueda/PopGenCode/blob/master/run10kdxy.pbs)), the script [dxy_wsfs.py](http://github.com/marqueda/PopGenCode/blob/master/dxy_wsfs.py) computes D<sub>xy</sub> for each line in the file.

## Cluster Separation Score (CSS) from Jones et al. 2012 Nature

The R-script [CSSm.R](http://github.com/marqueda/PopGenCode/blob/master/CSSm.R) computes the Cluster Separation Score (CSS) between two groups (clusters) of individuals as introduced by [Jones et al. 2012, Nature](https://doi.org/10.1038/nature10944). As the CSS formula in the supplementary material of Jones et al. 2012 contains an error, the script uses the corrected formula of [Miller et al. 2019, Current Biology](https://doi.org/10.1016/j.cub.2018.12.044), but without dividing the CSS score by the number of sites in a window (Cs' in Miller et al. 2019). The R-script takes a VCF file and a list of group (cluster) assignments for all individuals as input, allows to either use multi-dimensional scaling (MDS) with two factors or principal components analysis (PCA, PC1 and PC2) for the CSS compuation and to define windows either as fixed size or number of loci windows.

The R-script [CSSm_permutation.R](http://github.com/marqueda/PopGenCode/blob/master/CSSm_permutation.R) uses output from the CSSm.R script and re-computes CSS scores for  permuted group memberships, either randomly across all individuals or within pre-defined strata (e.g. within populations if several populations are included in an analysis and each population has members of both groups), and outputs empirical quantiles of the observed CSS values given the permuted CSS values for each window.
